framework/web/ErrorHandler.php

```diff
@@ -482,7 +482,7 @@ public function argumentsToString($args)
      */
     public function getExceptionName($exception)
     {
-        if ($exception instanceof \yii\base\Exception || $exception instanceof \yii\base\InvalidCallException || $exception instanceof \yii\base\InvalidParamException || $exception instanceof \yii\base\UnknownMethodException) {
+        if ($exception instanceof \yii\base\Exception || $exception instanceof \yii\base\InvalidCallException || $exception instanceof \yii\base\InvalidArgumentException || $exception instanceof \yii\base\UnknownMethodException) {
```

Comment:
I think we should leave it unchanged. We should also handle old `InvalidParamException` in this way - if `InvalidArgumentException` will extend `InvalidParamException` then `instanceof InvalidParamException` will handle both.

This also applies to many try-catch blocks, which may catch user exceptions.

Reply:
No it’s the other way around. `InvalidParamException` must extend `InvalidArgumentException`. Otherwise you would always be dealing with a deprecated class (either directly or via its parent).

Reply:
> Otherwise you would always be dealing with a deprecated class (either directly or via its parent).

This is the cost of BC. You can't change thrown exception in this way, because all `instanceof` or `catch` will stop working correctly.

Reply:
OK I see what you’re saying. I just reverted most of the `InvalidArgumentException` changes; now that class exists as a subclass of `InvalidParamException`, and `InvalidParamException` has been deprecated, so apps can choose to throw `InvalidArgumentException`s in their own code now, but Yii itself still only throws/catches/checks instanceof `YiiParamException` until 2.1.

Reply:
You misunderstood me. I still think that we should throw new `InvalidArgumentException` everywhere. This will not break BC since all `$exception instanceof InvalidParamException` and `catch (InvalidParamException $exception)` will work as before. But we need to use old `InvalidParamException` for `instanceof` and `catch` in framework, to make sure that old apps and extensions, which may throw old exception, will work as before. In this way users may start using  new exceptions everywhere (even for `instanceof` and `catch` for exceptions thrown by framework), and old code will still work.

Reply:
@rob006 Gotcha - OK now everything throws `InvalidArgumentException`s, but anything that catches `InvalidParamExceptions` or checks for `instanceof InvalidParamException` has been left alone.

Also I’ve left `@throws InvalidParamException` tags alone in `yii\db\conditions\ConditionInterface`, `yii\rbac\CheckAccessInterface`, and `yii\rbac\MessageInterface`, since changing them in an interface felt wrong. Let me know if you think that wasn’t the right call.